### PR TITLE
RA-1597: End Date validation for conditions

### PR DESCRIPTION
**Pull request by haripriya999**
# RA-1597: Validation of End Date of a Condition

## Description of what I changed
Validation of end date in add conditions

## Issue I worked on

see https://issues.openmrs.org/browse/RA-1597

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: omod/src/main/webapp/pages/conditionlist/addCondition.gsp [hunk 1]
```diff
@@ -43,15 +43,16 @@
 <div id="condition" ng-app="conditionApp" ng-controller="ConditionController">
     <h2 class="inline">${ui.message('coreapps.conditionui.addNewCondition')}</h2><br/>
 
-    <div class="horizontal">
-    <label>${ui.message('coreapps.conditionui.condition')} </label>
```
Comment (dkayiwa) [inline omod/src/main/webapp/pages/conditionlist/addCondition.gsp]: Any reason why we moved the label from the above line?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 2]
```diff
@@ -20,6 +20,10 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
      * @type {Function}
      */
     self.saveCondition = self.saveCondition || function () {
+        if($scope.condition.onSetDate>$scope.condition.endDate){
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Can you put some spaces around the > sign?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 3]
```diff
@@ -31,25 +35,46 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
         }
+    }
 
     self.initCondition = self.initCondition || function () {
             $scope.patientUuid = CommonFunctions.extractUrlArgs(window.location.search)['patientId'];
             $scope.condition = new ConditionModel($scope.patientUuid);
+            self.showEndDate();
         }
 
-    self.validateCondition = self.validateCondition || function () {
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Why was this removed?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 4]
```diff
@@ -31,25 +35,46 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
+        }
+        else{
+            groups[2].style.visibility = "hidden";    
+        }
+     }   
 
-            self.saveCondition();
+    self.validateCondition = self.validateCondition || function () {
+        var concept ;
+        if($scope.concept.concept.uuid)
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Have all these new changes got anything to do with the ticket?

File: omod/src/main/webapp/pages/conditionlist/addCondition.gsp (hunk 1 above)
Comment (haripriya999) [reply omod/src/main/webapp/pages/conditionlist/addCondition.gsp]: In order to get the condition , onset date and end date in same alignment.

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 3 above)
Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: This function is not being removed. It is being implemented after showEndDate function.

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 2 above)
Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Sure :)

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 4 above)
Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: It is to validate the information before the condition gets saved.

Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: As we also have to validate the end date for inactive conditions, I had to make those changes for validation of end date.

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 5]
```diff
@@ -31,25 +35,46 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
+            concept=$scope.condition.concept;
+        }
+        if (concept !== null) {
+            $scope.condition.concept = new ConceptModel(concept.uuid, concept.display);
+        } else {
+            // noncoded diagnosis
+            $scope.condition.conditionNonCoded = "NON_CODED:" + $scope.concept.word;
+        }
+        $scope.condition.onSetDate = self.getSelectedDate();
+        if($scope.condition.status=='INACTIVE')
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Why are we using 'INACTIVE' in more than one place?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 6]
```diff
@@ -31,25 +35,46 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
+        {
+            $scope.condition.endDate = self.getEndDate();
+        }
+        self.saveCondition();
+    }
 
-            self.saveCondition();
+    self.showEndDate = self.showEndDate || function() {
+        var groups = document.getElementsByClassName("group");
+        if($scope.condition.status=='INACTIVE'){
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Why do we repeat 'INACTIVE'?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 5 above)
Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Here I'm are checking if condition is INACTIVE or not, to save the value of end date for inactive conditions in the database

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 6 above)
Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Here I'm checking if condition status is 'INACTIVE', in order to make the end date field visible only when we choose the inactive radio button.

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 5 above)
Comment (dkayiwa) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: What happens when we change 'INACTIVE' to 'INNNNACTIVE'?

Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Why would $scope.condition.status become 'INNNNACTIVE' when condition.status takes only two values 'ACTIVE' and 'INACTIVE' ?

Comment (dkayiwa) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: In how many places would you have to change it?

Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Are you suggesting to use a constant?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 7]
```diff
@@ -9,7 +9,7 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
 
     $scope.patientUuid = null;
     $scope.condition = null;
-
+    const inactiveStatus = 'INACTIVE';
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Have you looked at our convention for constants? https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 5 above)
Comment (dkayiwa) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: What do you think?

**Comment by dkayiwa** [comment]
Can you also check the merge conflicts?

**Comment by haripriya999** [comment]
@dkayiwa  PR has been updated

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 8]
```diff
@@ -31,34 +36,62 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
+        }
+     }        
 
     self.getSelectedDate = self.getSelectedDate || function () {
             var datePicker = angular.element(document.getElementsByName('conditionStartDate'))[0];
             return datePicker.value;
         }
 
+    self.getEndDate = self.getEndDate || function() {
+            var endDatePicker = angular.element(document.getElementsByName('conditionEndDate'))[0];
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Do we need a temporary variable just for the sake of calling .value on it?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 9]
```diff
@@ -31,34 +36,62 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
+        if($scope.concept.concept.uuid) {
+            concept=$scope.concept.concept;
+        } else {
+            concept=$scope.condition.concept;
+        }
+
+        if (concept !== null) {
+            $scope.condition.concept = new ConceptModel(concept.uuid, concept.display);
+        } else {
+            // noncoded diagnosis
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: What is the use of the above comment?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 10]
```diff
@@ -31,34 +36,62 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
-            var concept = $scope.concept.concept;
-            if (concept !== null) {
-                $scope.condition.concept = new ConceptModel(concept.uuid, concept.display);
-            } else {
-                // noncoded diagnosis
-                $scope.condition.conditionNonCoded = "NON_CODED:" + $scope.concept.word;
-            }
+        var concept ;
+        if($scope.concept.concept.uuid) {
+            concept=$scope.concept.concept;
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Can you always put space around the = sign?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js [hunk 11]
```diff
@@ -31,34 +36,62 @@ function ConditionController($scope, RestfulService, ConditionModel, ConceptMode
...
+        } else {
+            // noncoded diagnosis
+            $scope.condition.conditionNonCoded = "NON_CODED:" + $scope.concept.word;
+        }
 
-            $scope.condition.onSetDate = self.getSelectedDate();
+        $scope.condition.onSetDate = self.getSelectedDate();
 
-            self.saveCondition();
+        if($scope.condition.status==INACTIVE_STATUS) {
```
Comment (dkayiwa) [inline omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: Can you put space around the == sign?

**Comment by dkayiwa** [comment]
Did you claim the ticket?

File: omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js (hunk 8 above)
Comment (haripriya999) [reply omod/src/main/webapp/resources/scripts/conditionlist/controllers/addcondition.controller.js]: I think it's more clear to write it this way. Also getSelectedDate()  has been written in the same fashion

**Comment by haripriya999** [comment]
@dkayiwa I updated the PR and also claimed the ticket
